drivers: gpio: miwu: Add miwu and gpio drivers in npcx7 series.

MulinChao opened the pull request:
Add Multi-Input Wake-Up Unit (MIWU) and GPIO driver implementations in Nuvoton npcx7 series..

It is separated into three commits. First, commit introduces MIWU which provides signal conditioning such as 'Level' and 'Edge' trigger type and grouping of NVIC's external interrupt sources. It's crucial since many devices such as gpio, timer, eSPI VW signals relies on it for interrupt handling. Beside gpio driver implementation, the other commits also demonstrate how we handle wake-up signals of gpio and uart by MIWU api functions.

MulinChao committed:
driver: intc: add MIWU driver support in NPCX series.

The device Multi-Input Wake-Up Unit (MIWU) supports the embedded
controller (EC) to exit 'Sleep' or 'Deep Sleep' power state which allows
chip has better power consumption. Also, it provides signal conditioning
such as 'Level' and 'Edge' trigger type and grouping of external
interrupt sources of NVIC. The NPCX series has three identical MIWU
modules: MIWU0, MIWU1, MIWU2. Together, they support a total of over 140
internal and/or external wake-up sources.

In this CL, we use device tree files to present the relationship bewteen
MIWU and the other devices in different npcx series. For npcx7 series,
it include:
1. npcx7-miwus-int-map.dtsi: it presents relationship between MIWU group
   and NVIC interrupt in npcx7. Please notice it isn't 1-to-1 mapping.
2. npcx7-miwus-wui-map.dtsi: it presents relationship between input of
   MIWU and its source device such as gpio, timer, eSPI VWs and so on.

This CL also includes:
1. Add MIWU device tree declarations.
2. MIWU api function declarations and implementation to configure signal
   conditions and callback function mechanism. They can be be classified
   into two types. One is for GPIO which connects original gpio callback
   implemetation and the other is for generic devices such as timer,
   eSPI, and so on.

Signed-off-by: Mulin Chao <MLChao@nuvoton.com>

MulinChao committed:
driver: gpio: add gpio driver support in NPCX series

Add gpio support for Nuvoton NPCX series. This CL includes:

1. Add GPIO device tree declarations.
2. Introduce wui_maps property in yaml file to present relationship
   between Wake-Up
   Input (WUI) and 8 IOs belong to the device.
3. Zephyr GPIO api implementation.
4. GPIO callback functions implementation with MIWU api functions.
5. Overlay file for gpio basic tests

Signed-off-by: Mulin Chao <MLChao@nuvoton.com>

MulinChao committed:
driver: uart: add wake-up support for uart device.

This CL configures the UART wake-up event triggered from a falling edge
(START condition) on CR_SIN pin. It also includes:

1. Introduce wui_maps property in yaml file to present relationship
   between Wake-Up Input (WUI) and UART device.
2. Implement wake-up mechanism by MIWU api functions.

Signed-off-by: Mulin Chao <MLChao@nuvoton.com>

MulinChao commented:
From the build error report, it shows "gen_isr_tables.py: error: multiple registrations at table_index 61 for irq 61 (0x3d)".  Dig the source of gen_isr_table, the root cause is that the test suit also uses IRQ61 (https://github.com/zephyrproject-rtos/zephyr/blob/master/tests/kernel/gen_isr_table/src/main.c#L283) which is the same as miwu2 group2 interrupt. 
Hi @ioannisg 
Might I know what is the purpose of the test suite? Is the workaround moving TEST_NUM_IRQS to unused IRQ? Thanks.

MulinChao committed:
tests: kernel: gen_isr_table: Add workaround for npcx7m6fb soc.

In NPCX7M6FB, it uses some the IRQs at the end of the vector table,
for example, the irq 60 and 61 used for Multi-Input Wake-Up Unit (MIWU)
device by default, and conflicts with isr used for testing. Moving IRQs
for this test suite to solve the issue.

Signed-off-by: Mulin Chao <MLChao@nuvoton.com>

MulinChao commented on drivers/interrupt_controller/intc_miwu.c:
```diff
@@ -0,0 +1,321 @@
+#include <logging/log.h>
+LOG_MODULE_REGISTER(intc_miwu, LOG_LEVEL_ERR);
+
+/* MIWU module instances declaration */
```
Done.

MulinChao commented on drivers/interrupt_controller/intc_miwu.c:
```diff
@@ -0,0 +1,321 @@
+
+		/*
+		 * Clear pending bit since it might be set
+		 * if WKINEN bit is changed.
```
Done.

MulinChao commented on drivers/interrupt_controller/intc_miwu.c:
```diff
@@ -0,0 +1,321 @@
+	static int intc_miwu_init##inst(struct device *dev)
+
+/* MIWU ISR implementation */
+#define NPCX_MIWU_ISR_FUNC_IMPL(inst)                                          \
```
Yes, we need '#define' since it relies on device tree files. I comment on what it connects ISR and MIWU group IRQ at the top of the driver.

MulinChao commented on drivers/interrupt_controller/intc_miwu.c:
```diff
@@ -0,0 +1,321 @@
+		} while (grp_mask != 0);                                       \
+	}
+
+/* MIWU init function implementation */
```
Done. I comment on what it does at the top of the driver.

MulinChao commented on dts/arm/nuvoton/npcx7m6fb.dtsi:
```diff
@@ -10,6 +10,10 @@
 
 /* NPCX7 series pinmux mapping table */
 #include "npcx/npcx7-alts-map.dtsi"
+/* NPCX7 series mapping table between MIWU wui bits and source device */
```
Done. I comment on what's MIWU and WUI at the top of the driver file.

MulinChao commented on soc/arm/nuvoton_npcx/common/reg/reg_def.h:
```diff
@@ -404,4 +404,26 @@ struct miwu_reg_t {
+ */
+struct gpio_reg_t {
+	/* 0x000: Port GPIOx Data Out */
+	volatile uint8_t PDOUT;
```
Regarding the lower case, could we still keep npcx register names as an upper case? I'd like to keep them the same as our datasheet possibly.  And we usually use a variable with a lower case to present it's a variable in ram.

sjg20 replied:
My understanding is that upper case is for #defines. It is pretty common to use lower case even if datasheets don't.

Don't you see a checkpatch warning with this?

MulinChao replied:
No, I don't. It passed  checkpatch.pl and got no warnings on my side. And the previous PR also.

MulinChao commented on drivers/serial/uart_npcx.c:
```diff
@@ -334,6 +338,19 @@ static int uart_npcx_init(struct device *dev)
 	config->uconf.irq_config_func(dev);
 #endif
 
+#ifdef CONFIG_SYS_POWER_DEEP_SLEEP_STATES
```
Done.

MulinChao commented on tests/kernel/gen_isr_table/src/main.c:
```diff
@@ -50,6 +50,13 @@ extern uint32_t _irq_vector_table[];
  */
 #define TEST_NUM_IRQS	30
+#elif defined(CONFIG_SOC_NPCX7M6FB)
+/* In NPCX7M6FB, it uses some of the IRQs at the end of the vector table, for
```
Done.

MulinChao commented on tests/kernel/gen_isr_table/src/main.c:
```diff
@@ -50,6 +50,13 @@ extern uint32_t _irq_vector_table[];
+#elif defined(CONFIG_SOC_NPCX7M6FB)
+/* In NPCX7M6FB, it uses some of the IRQs at the end of the vector table, for
+ * example, the irq 60 and 61 used for Multi-Input Wake-Up Unit (MIWU) device
+ * by default, and conflicts with isr used for testing. Moving IRQs for this
```
Done. The other approach is setting all miwu devices statuses in device tree as 'disabled'. It won't install ISRs such as IRQ 60 and 61 by default in this test suite. But you need to enable them in a board device tree file. For example, you need to add the following declarations in npcx7m6fb.dts file.

_&miwu0 {
	status = "okay";
};
&miwu1 {
	status = "okay";
};
&miwu2 {
	status = "okay";
};_

It's crucial to enable all MIWU devices by default since lots of devices' interrupt handler relies on them. Adding them in all boards' device tree files use npcx ec is necessary.

MulinChao commented on soc/arm/nuvoton_npcx/common/soc_gpio.h:
```diff
@@ -0,0 +1,31 @@
+ * @retval Pointer to structure device
+ * @retval NULL Invalid parameter of GPIO port index
+ */
+struct device *soc_get_gpio_dev(int port);
```
I have thought of using device_get_binding() to solve it. But we need to maintain a gpio device name array in miwu driver or using sprintf() to generate gpio device name string to get device address dynamically with port parameter. They might not be better than the current solution. That's all I can figure out so far.

sjg20 replied:
Right, but I am wondering why you need the port parameter. Typically if you have a device for each port, then you can get the device from the DT and the idea of a port number is no-longer userful.

Anyway if you can't figure it out, let's leave it as is for now.

MulinChao replied:
Regarding why we need the 'port' parameter, this is because the original Zephyr gpio callback mechanism restores the gpio callback structures in the heap directly. Please refer https://github.com/zephyrproject-rtos/zephyr/blob/master/include/drivers/gpio.h#L363 and https://github.com/zephyrproject-rtos/zephyr/blob/master/samples/basic/button/src/main.c#L41. Since they are declared and registered by gpio_add_callback() in the application layer, the npcx miwu driver has no chance to know how many gpios needs the MIWU's support during compiling. (Please notice a gpio event triggers related WUI signal and invokes MIWU's ISR.) 

Hence, we handle gpio callback functions in miwu driver directly. By borrowing unused 24 bits of pin_mask which is part of gpio_callback structure and already exists in heap, the gpio driver saved the necessary information in manage_callback() api function. It includes:
1. Related WUI signal of target gpio which is used for setting its MIWU device configuration.
2. Port number of this gpio which is used for getting its gpio device address during invoking its callback functions in MIWU's ISR. Since there is only unused 24 bits space, we cannot save the device address directly. Saving port number is an alternative solution.

sjg20 replied:
OK I see. I am not enough of an expert to know if there is a better way, so it seems OK to me.

MulinChao commented:
Hi @sjg20 : 
I left two comments for further discussion and resolved the others in this PR. Putting a comment at the top of miwu driver is a good idea since it's not a popular hardware device. Thanks for reviewing it.

pabigot reviewed (commented):
Looks interesting, but there's no public information I can find on the board or its MCU, or where to buy the board.

jettr reviewed (commented):
Sorry, I didn't have a change to finish reviewing. I just a few minor comments. Looks good to me so far though

pabigot commented on drivers/gpio/gpio_npcx.c:
```diff
@@ -0,0 +1,298 @@
+	}
+
+	/* Don't support "open source" mode */
+	if (((flags & GPIO_SINGLE_ENDED) != 0U) &&
```
There's a strong push in Zephyr to use MISRA compliant practices, which mandates a strict essential type model: no implicit type conversions.  Boolean is a MISRA essential type; bitwise operators apply to and produce values that are essentially unsigned type.

This is rule 14.4 in https://docs.zephyrproject.org/latest/guides/coding_guidelines/index.html#main-rules: 

> The controlling expression of an if statement and the controlling expression of an iteration-statement shall have essentially Boolean type

The correct style for this is `(flags & FLAG) != 0` over `!!(flags & FLAG)` or `(flags & FLAG)` even if the language specifies they're all equivalent.

Many of the uses near this one are violations, but this one is not.

Zephyr isn't enforcing this currently, but it will happen someday.

jettr replied:
That you for the background, Peter! Let's go with `(flags & FLAG) != 0` style then in all cases in this PR so we have less to change in the future.

MulinChao replied:
Done.

MulinChao commented on drivers/gpio/gpio_npcx.c:
```diff
@@ -0,0 +1,298 @@
+	/* gpio_driver_config needs to be first */
+	struct gpio_driver_config common;
+	/* GPIO controller base address */
+	uint32_t base;
```
I have tried it before. It will cause a warning during code style checking. Please see below.

_-:9: WARNING:NEW_TYPEDEFS: do not add new typedefs
#9: FILE: drivers/gpio/gpio_npcx.c:55:
+typedef struct gpio_reg gpio_reg_t;

- total: 0 errors, 1 warnings, 7 lines checked_

jettr replied:
I think the static cast in the initialization structure make the most sense because you are trying to format the data in the best representation there. I would rather see the cast once in initialization rather than having the cast every time we call it. Also specifying the datatype here in the structure makes it easier to know what type of data to expect.

andrewboie replied:
really would prefer you use `uintptr_t` instead of `uint32_t` for linear addresses treated as integral values

MulinChao replied:
Ok. I will add one more commit here to use uintptr_t instead of uint32_t including the previous PR.

MulinChao replied:
Done.

MulinChao commented:
> Looks interesting, but there's no public information I can find on the board or its MCU, or where to buy the board.

Hi @pabigot:
Thanks for your comments in this PR. Regarding the npcx board, our PM will contact you privately.

Hi @michalsch:
Could you help with this? Thanks.

MulinChao commented on drivers/gpio/gpio_npcx.c:
```diff
@@ -0,0 +1,298 @@
+static int gpio_npcx_config(struct device *dev,
+			     gpio_pin_t pin, gpio_flags_t flags)
+{
+	struct gpio_reg_t *inst = HAL_INSTANCE(dev);
```
> As far as I have seen in this codebase, the `_t` suffix is used mainly with typedefs
> 
> So you would have
> 
> ```
> struct gpio_reg;
> typedef struct gpio_reg gpio_reg_t;
> ```

That's a good idea. I will add one more commit in this PR to change all 'reg_t' to 'reg'.

MulinChao replied:
Done.

MulinChao commented on drivers/gpio/gpio_npcx.c:
```diff
@@ -0,0 +1,298 @@
+int gpio_npcx_init(struct device *dev)
+{
+	const struct gpio_npcx_config *config = DRV_CONFIG(dev);
+	ARG_UNUSED(config);
```
config is used for __ASSERT macro function. If I drop "ARG_UNUSED(config)", it cannot pass checkpatch.pl.

pabigot replied:
But `config` isn't an argument, so this is wrong (or at least misleading).

I usually do an explicit `(void)config;` right before the `__ASSERT` to provide a clue.  I think that's also something MISRA considers a violation, but it works for now.

Or you could do `ARG_UNUSED(dev);` then use `DRV_CONFIG(dev)->wui_size` in the assert.

MulinChao replied:
Thanks for your suggestions. I picked the latter and updated this PR.

MulinChao commented on drivers/gpio/gpio_npcx.c:
```diff
@@ -0,0 +1,298 @@
+	struct gpio_reg_t *inst = HAL_INSTANCE(dev);
+
+	/* Toggle raw bits of GPIO output registers */
+	inst->PDOUT ^= mask;
```
Since it has been discussed in the other comment, I resolve it now.

MulinChao commented on drivers/gpio/gpio_npcx.c:
```diff
@@ -0,0 +1,298 @@
+}
+
+static int gpio_npcx_port_toggle_bits(struct device *dev,
+						gpio_port_value_t mask)
```
It will exceed 80 characters if I put them on the same line.

MulinChao commented on drivers/gpio/gpio_npcx.c:
```diff
@@ -0,0 +1,298 @@
+
+/* Driver convenience defines */
+#define DRV_CONFIG(dev) \
+	((const struct gpio_npcx_config *)(dev)->config)
```
Done.

MulinChao commented on dts/bindings/interrupt-controller/nuvoton,npcx-miwu-int-map.yaml:
```diff
@@ -0,0 +1,32 @@
+        description: parent device node of miwu groups
+
+child-binding:
+    description: Child node to present the mapping between MIWU group and interrupt
```
Done

MulinChao commented on dts/bindings/interrupt-controller/nuvoton,npcx-miwu-int-map.yaml:
```diff
@@ -0,0 +1,32 @@
+       irq_prio:
+          type: int
+          required: true
+          description: irq's priority for miwu group
```
Done.

MulinChao commented on soc/arm/nuvoton_npcx/common/soc_dt.h:
```diff
@@ -42,5 +42,61 @@
+	DT_INST_PHANDLE(inst, wui_map)
+
+/* Construct a npcx_wui structure from wui_map property */
+#define DT_NPCX_WUI_ITEM(inst)				                       \
```
Done.

MulinChao commented on soc/arm/nuvoton_npcx/common/soc_miwu.h:
```diff
@@ -0,0 +1,207 @@
+ *
+ * Used to indicate a Muti-Input Wake-Up Unit (MIWU) source belongs to which
+ * group and bit of MIWU modules.
+ */
```
Done.

MulinChao commented on soc/arm/nuvoton_npcx/common/soc_miwu.h:
```diff
@@ -0,0 +1,207 @@
+ * Beware the size of such structure must equal struct gpio_port_pins_t.
+ */
+struct miwu_io_params {
+	uint8_t pin_mask; /** A mask of pins the callback is interested in. */
```
I  referred to code style here. (https://github.com/zephyrproject-rtos/zephyr/blob/master/include/drivers/gpio.h#L367)  I think it's OK.

MulinChao commented on soc/arm/nuvoton_npcx/npcx7/soc.c:
```diff
@@ -42,3 +42,9 @@ NPCX_REG_SIZE_CHECK(uart_reg_t, 0x027);
+/* MIWU register structure check */
+NPCX_REG_SIZE_CHECK(miwu_reg_t, 0x078);
+NPCX_REG_OFFSET_CHECK(miwu_reg_t, LOW_PND_WU, 0x00B);
+NPCX_REG_OFFSET_CHECK(miwu_reg_t, WKPND6, 0x02E);
```
Regarding lower case for hex value, I found lots of examples use upper case for hex value. Is it a requirement for code style?

pabigot replied:
I don't think so (the coding style doesn't go to that level of detail).  I would probably use upper case for small constants like this, but lower case in larger values because it's more readable.

What's a bit odd here is the leading zero making it a 12-bit constant rather than an 8-bit one.  Is that intentional?

MulinChao commented on soc/arm/nuvoton_npcx/common/reg/reg_def.h:
```diff
@@ -258,4 +258,150 @@ struct uart_reg_t {
+};
+
+/* MIWU multi-registers */
+#define NPCX_WKEDG_OFFSET(n)    (0x000 + ((n) * 2L) + ((n) < 5 ? 0 : 0x1E))
```
The other driver might access these registers such as WKINEN, WKEN, and so on for the other purpose directly. I suggest composing the other PR after all npcx drivers are landed. Then we can have an overall perspective to do refactoring.

jettr replied:
I think this is okay to land now. We could come back and refactor this if it makes sense. (I do think inline functions could be nice as well)

MulinChao committed:
drivers: npcx: remove suffix '_t' of device register structure.

Remove the '_t' suffix of device register structure since it is used
mainly with typedefs.

Signed-off-by: Mulin Chao <MLChao@nuvoton.com>

pabigot commented:
> > Looks interesting, but there's no public information I can find on the board or its MCU, or where to buy the board.
> 
> Hi @pabigot:
> Thanks for your comments in this PR. Regarding the npcx board, our PM will contact you privately.

Thanks.  But my main point is an ongoing concern with adopting into mainline Zephyr hardware (boards, SoCs, shields) that are not generally available.  This introduces a near-permanent maintenance burden as tree-wide refactoring often requires their code be changed, but the people making those changes can't easily verify functionality.  I really hate changing something when I can't verify I didn't break it; so I have 26 Zephyr target boards spanning almost all SoC families.  More than half of those are not of interest to the people who fund most of my Zephyr work.

When the documentation is also not available it becomes a bigger problem.

Note that this is my personal view, which others may or may not share.  There is no documented policy in Zephyr providing non-technical criteria for adding new content.  Since this target has already been accepted clearly the new peripheral implementations must go in.  But the level of review I can give is seriously impeded by lack of technical documentation and the ability to test.

jettr commented:
> Thanks. But my main point is an ongoing concern with adopting into mainline Zephyr hardware (boards, SoCs, shields) that are not generally available. This introduces a near-permanent maintenance burden as tree-wide refactoring often requires their code be changed, but the people making those changes can't easily verify functionality. I really hate changing something when I can't verify I didn't break it; so I have 26 Zephyr target boards spanning almost all SoC families. More than half of those are not of interest to the people who fund most of my Zephyr work.
> 
> When the documentation is also not available it becomes a bigger problem.
> 
> Note that this is my personal view, which others may or may not share. There is no documented policy in Zephyr providing non-technical criteria for adding new content. Since this target has already been accepted clearly the new peripheral implementations must go in. But the level of review I can give is seriously impeded by lack of technical documentation and the ability to test.

I think your concern about testing wide-scale changes/refactors is valid. I know I have heard something about a hardware test matrix that we can run tests on that is distributed out to different chip vendors; would adding more testing for the npcx product line address some of your concerns here? @nashif for more details on test matix.

MulinChao commented:
@pabigot and @jett:
Regarding test suites, boards, and documentation during refactoring, I fully understand what you concern. Each time I did changes for comments from code review, passing the related driver test suite is necessary before I submitting them. I think it's OK to provide npcx7m6fb evaluation boards. 
Regarding the document or the datasheet of Nuvoton EC, I need further discussion with our PM. (Basically, they are provided under NDA.)

nashif commented:
@jettr @pabigot

> Thanks. But my main point is an ongoing concern with adopting into mainline Zephyr hardware (boards, SoCs, shields) that are not generally available.

This was already discussed in multiple occasions and the consensus was that support for unreleased hardware and missing documentation of the hardware is acceptable as long ad the drivers, SoCs and boards are being actively maintained with someone available to answer any technical questions. If this changes over time and some hardware becomes unmaintained and it prevents us from moving forward because of lack of response or broken support, we will then reserve the right to disable or remove such support.

Supporting hardware that is still unreleased is often related to product development and is an indicator that the project is doing well in general :) Linux is full of drivers and hardware support that is still not released, there is no way we can make progress only with release hardware (that is maybe how we started 5 years ago, but we are way beyond that).

jettr commented on drivers/gpio/gpio_npcx.c:
```diff
@@ -0,0 +1,294 @@
+	int pin = find_lsb_set(callback->pin_mask) - 1;
+
+	/* Has the IO pin valid MIWU input source? */
+	if (config->wui_maps[pin].table == NPCX_MIWU_TABLE_NONE) {
```
should we validate that pin != -1 before accessing into array? What guarantee do we have the callback->pin_mask is not 0? (If we do have those guarantees, then we should call them out in comments here)

MulinChao replied:
Done.

jettr commented on drivers/interrupt_controller/intc_miwu.c:
```diff
@@ -0,0 +1,368 @@
+	callback->source = source;
+}
+
+int soc_miwu_manage_gpio_callback(struct miwu_io_callback *cb, bool set)
```
Why can't we use `gpio_manage_callback` in gpio_utils.h here instead

MulinChao replied:
First, the parameter is different between soc_miwu_manage_gpio_callback() and gpio_manage_callback() although 'struct miwu_io_callback *cb' and struct gpio_callback *callback share the same memory space. In this PR, it won't increase code size since we don't call any gpio_manage_callback() and replace it with soc_miwu_manage_gpio_callback(). 

Of course, we can cast 'struct miwu_io_callback *'  back to 'struct gpio_callback * and call gpio_manage_callback() in MIWU driver directly. But I thought the current one is more straightforward.

jettr commented on drivers/interrupt_controller/intc_miwu.c:
```diff
@@ -0,0 +1,368 @@
+	return 0;
+}
+
+int soc_miwu_manage_dev_callback(struct miwu_dev_callback *cb, bool set)
```
same, seems like we should be using `gpio_manage_callback` instead of implementing it again

MulinChao replied:
No, we cannot use gpio_manage_callback directly since their parameters are not the same. But we can abstract the same mechanism to "int miwu_manage_callback(sys_slist_t *callbacks, struct sys_snode_t *node, bool set)". Then call it in soc_miwu_manage_dev_callback() and int soc_miwu_manage_gpio_callback() separately, What do you think?

jettr replied:
Ah, okay. I missed the type difference. It would be nice to single source this, but the type-safety is more important. Closing issue.

jettr commented on drivers/interrupt_controller/intc_miwu.c:
```diff
@@ -0,0 +1,368 @@
+/* Driver config */
+struct intc_miwu_config {
+	/* miwu controller base address */
+	uint32_t base;
```
Similar to other comments, this uint32_t type would be better as a uintprt_t or a specialized typedef

MulinChao replied:
Done.

jettr commented on drivers/interrupt_controller/intc_miwu.c:
```diff
@@ -0,0 +1,368 @@
+			cb->handler(soc_get_gpio_dev(cb->params.gpio_port),
+					(struct gpio_callback *)cb,
+					cb->params.pin_mask);
+		}
```
Do you expect more than one call back to match? If not, let's break. If so, let's add a comment saying that multiple GPIO callback may match

MulinChao replied:
Done. I think it is not one to one relationship and add more comment here.

jettr commented on drivers/interrupt_controller/intc_miwu.c:
```diff
@@ -0,0 +1,368 @@
+				&& cb->wui.bit == wui_bit) {
+			__ASSERT(cb->handler, "No Generic callback handler!");
+			/* Execute generic callback function */
+			cb->handler(cb->source, &cb->wui);
```
Same, either break or add comment about multiple matching please

MulinChao replied:
Done.

jettr commented on soc/arm/nuvoton_npcx/common/reg/reg_def.h:
```diff
@@ -258,4 +258,172 @@ struct uart_reg_t {
+ */
+struct miwu_reg {
+	/* 0x000: Edge Detection 1 */
+	volatile uint8_t WKEDG1[1];
```
More curious than anything: why are these arrays of size 1 instead of just non-array values?

(I am not asking you to change; I just want to learn the reason)

MulinChao replied:
At the beginning, the NPCX_WKEDG macro function is "NPCX_WKEDG_OFFSET(inst, n)    (inst->WKEDG1 + ((n) * 2L) + ((n) < 5 ? 0 : 0x1E))" Since we have use the offset directly in previous PR, I left them in case we need to access LOW_PND_WU register. The other registers can be replaced with macro functions.

jettr replied:
If we aren't using this, then let's remove it

MulinChao replied:
Done.

jettr commented on drivers/gpio/gpio_npcx.c:
```diff
@@ -0,0 +1,298 @@
+
+/* Driver config */
+struct gpio_npcx_config {
+	/* gpio_driver_config needs to be first */
```
+1; there do seem to still by /* instead of /** here

MulinChao replied:
Oops. I missed it. Sorry about that. I will use 'Files changed' tab to browse all comments.
Per my understanding, a '/**' prefix locates at the beginning of doxygen format comments. If so,  I prefer the prefix here is '/*'.

jettr reviewed (commented):
The remaining comments are relatively minor. This PR looks good to me overall! Can you make a sweep through all of the remaining open comments and address them, please?

MulinChao committed:
drivers: npcx: replace type of register base address with uintptr_t

Replace npcx register base address type, uint32_t, with uintptr_t.
It is easier to know what type of base address and for linear
addresses treated as integral values.

This CL also modified IS_BIT_SET() macro function to fit MISRA code
guidelines.

Signed-off-by: Mulin Chao <MLChao@nuvoton.com>

MulinChao commented:
> Can you make a sweep through all of the remaining open comments and address them, please?

@jettr I have swept all remaining comments. Most of them are resolved and related changes were updated. But a few comments are still open and I gave replies for further discussions. Please check them, thanks.

jettr commented on drivers/gpio/gpio_npcx.c:
```diff
@@ -0,0 +1,298 @@
+	 * after setting all other attributes, so as not to create a
+	 * temporary incorrect logic state 0:input 1:output
+	 */
+	if (!(flags & GPIO_OUTPUT))
```
Can you change these flags to command against `!= 0` as well in this function to remain consistent.

MaureenHelm replied:
@jettr It's hard for the release team to know whether to merge a PR when you simultaneously approve it and ask for changes in a comment.

MulinChao replied:
Done. I also modified IS_BIT_SET() macro function to fit MISRA coding guideline in the last commit.

jettr replied:
@MaureenHelm thanks for the feedback. I was using the process of giving an approval with the knowledge that the author will perform the small changes before submitting code. I forget that authors don't land the changes once they have approval. (I only do this when I also trust the author to do it correctly as well). I am still getting used to the GitHub workflow.

I will wait in the future for all changes to be made (even if small) before marking the approval. Thanks!

MaureenHelm replied:
No worries, thanks

jettr reviewed (approved):
Looks good!

Michal-Sch commented:
> @jettr @pabigot
> 
> > Thanks. But my main point is an ongoing concern with adopting into mainline Zephyr hardware (boards, SoCs, shields) that are not generally available.
> 
> This was already discussed in multiple occasions and the consensus was that support for unreleased hardware and missing documentation of the hardware is acceptable as long ad the drivers, SoCs and boards are being actively maintained with someone available to answer any technical questions. If this changes over time and some hardware becomes unmaintained and it prevents us from moving forward because of lack of response or broken support, we will then reserve the right to disable or remove such support.
> 
> Supporting hardware that is still unreleased is often related to product development and is an indicator that the project is doing well in general :) Linux is full of drivers and hardware support that is still not released, there is no way we can make progress only with release hardware (that is maybe how we started 5 years ago, but we are way beyond that).

Sorry for the delay. Was OOO and had to catch-up.
Datasheet can be provided under NDA as ML mentioned.
We can provide boards but I assume they are meaningless without the datasheet.
